cpp/src/tsne/exact_kernels.cuh

```diff
@@ -308,17 +325,18 @@ __global__ void repulsive_kernel_2d(
 }
 
 /****************************************/
-template <int TPB_X = 32, int TPB_Y = 32>
-float repulsive_forces(const float *restrict Y, float *restrict repel,
-                       const float *restrict norm, float *restrict Z_sum,
-                       const int n, const int dim,
-                       const float df_power,  // -(df + 1)/2)
-                       const float recp_df, cudaStream_t stream) {
-  CUDA_CHECK(cudaMemsetAsync(Z_sum, 0, sizeof(float) * 2 * n, stream));
-  CUDA_CHECK(cudaMemsetAsync(repel, 0, sizeof(float) * n * dim, stream));
+template <typename value_idx, typename value_t, int TPB_X = 32, int TPB_Y = 32>
+value_t repulsive_forces(const value_t *restrict Y, value_t *restrict repel,
+                         const value_t *restrict norm, value_t *restrict Z_sum,
+                         const value_idx n, const value_idx dim,
+                         const value_t df_power,  // -(df + 1)/2)
```

Comment:
We should keep the non-pointer arguments as 32-bit across all of these functions. This is going to affect performance and I'm not sure there is a benefit. The exception here would be any params related to representing the size, such as `n`.